### PR TITLE
Add `median`-based h2o queries

### PR DESCRIPTION
**Pull request by ncclementi**
- [x] Closes https://github.com/coiled/h2o-benchmarks/issues/5

@jrbourbeau I'm realizing the issue about this was opened in the h2o benchmark. Was that meant to be that way?

### CODE REVIEW

File: tests/benchmarks/h2o/test_h2o_benchmarks.py [hunk 1]
```diff
@@ -91,6 +93,19 @@ def test_q5(ddf, small_client):
     )
 
 
+@pytest.mark.skipif(
+    Version(dask.__version__) < Version("2022.10.0"),
+    reason="No support for median in dask < 2022.10.0",
+)
```
Comment (jrbourbeau) [inline tests/benchmarks/h2o/test_h2o_benchmarks.py:99]: Nice -- thanks for adding this!

**Review by jrbourbeau — COMMENTED** [review]
Thanks @ncclementi! 

> I'm realizing the issue about this was opened in the h2o benchmark. Was that meant to be that way?

I think adding the query both here, and to the notebook in the `h2o-benchmarks` repo would be good

File: tests/benchmarks/h2o/test_h2o_benchmarks.py [hunk 2]
```diff
@@ -91,6 +93,19 @@ def test_q5(ddf, small_client):
...
+def test_q6(ddf, small_client):
+    ddf = ddf[["id4", "id5", "v3"]]
+    (
+        ddf.groupby(["id4", "id5"], dropna=False, observed=True)
+        .agg({"v3": ["median", "std"]}, shuffle="tasks")
+        .compute()  # requires shuffle="tasks"
```
[the remark this answers is not in the document]
Comment (ncclementi) [reply tests/benchmarks/h2o/test_h2o_benchmarks.py:105]: I didn't specify it and I get a ValueError:
```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [7], line 4
      1 ddf = ddf[["id4", "id5", "v3"]]
      2 (
      3     ddf.groupby(['id4','id5'], dropna=False, observed=True)
----> 4     .agg({'v3': ['median','std']}).compute()
      5 )

File ~/mambaforge/envs/runtime-dashboard/lib/python3.10/site-packages/dask/dataframe/groupby.py:2514, in DataFrameGroupBy.agg(self, arg, split_every, split_out, shuffle)
   2512 @_aggregate_docstring(based_on="pd.core.groupby.DataFrameGroupBy.agg")
   2513 def agg(self, arg, split_every=None, split_out=1, shuffle=None):
-> 2514     return self.aggregate(
   2515         arg, split_every=split_every, split_out=split_out, shuffle=shuffle
   2516     )

File ~/mambaforge/envs/runtime-dashboard/lib/python3.10/site-packages/dask/dataframe/groupby.py:2508, in DataFrameGroupBy.aggregate(self, arg, split_every, split_out, shuffle)
   2505 if arg == "size":
   2506     return self.size()
-> 2508 return super().aggregate(
   2509     arg, split_every=split_every, split_out=split_out, shuffle=shuffle
   2510 )

File ~/mambaforge/envs/runtime-dashboard/lib/python3.10/site-packages/dask/dataframe/groupby.py:1976, in _GroupBy.aggregate(self, arg, split_every, split_out, shuffle)
   1974 has_median = any(s[1] in ("median", np.median) for s in spec)
   1975 if has_median and not shuffle:
-> 1976     raise ValueError(
   1977         "In order to aggregate with 'median', you must use shuffling-based "
   1978         "aggregation (e.g., shuffle='tasks')"
   1979     )
   1981 if shuffle:
   1982     # Shuffle-based aggregation
   1983     #
   (...)
   1988     # If we have a median in the spec, we cannot do an initial
   1989     # aggregation.
   1990     if has_median:

ValueError: In order to aggregate with 'median', you must use shuffling-based aggregation (e.g., shuffle='tasks')
```

[the remark this answers is not in the document]
Comment (ian-r-rose) [reply tests/benchmarks/h2o/test_h2o_benchmarks.py:105]: It is not intended to be auto-selected: ACA-based aggregations are still default for aggspecs that don't include `median`, and I wanted to be explicit about what adding `median` entails. So this looks correct to me.

**Comment by jrbourbeau** [comment]
We're getting the following error in the [`upstream` Python 3.9 build](https://github.com/coiled/coiled-runtime/actions/runs/3283632529/jobs/5408595119)

```python
____________________ ERROR at setup of test_dataframe_align ____________________
[gw0] linux -- Python 3.9.13 /usr/share/miniconda3/envs/test/bin/python

request = <SubRequest 'small_cluster' for <Function test_dataframe_align>>
dask_env_variables = {'DASK_COILED__TOKEN': '***'}

    @pytest.fixture(scope="module")
    def small_cluster(request, dask_env_variables):
        # Extract `backend_options` for cluster from `backend_options` markers
        backend_options = merge(
            m.kwargs for m in request.node.iter_markers(name="backend_options")
        )
        backend_options["send_prometheus_metrics"] = True
    
        module = os.path.basename(request.fspath).split(".")[0]
>       with Cluster(
            name=f"{module}-{uuid.uuid4().hex[:8]}",
            n_workers=10,
            worker_vm_types=["t3.large"],  # 2CPU, 8GiB
            scheduler_vm_types=["t3.xlarge"],
            backend_options=backend_options,
            package_sync=True,
            environ=dask_env_variables,
        ) as cluster:

conftest.py:439: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ClusterBeta(test_dataframe-9160c61b, '<Not Connected>', workers=0, threads=0, memory=0 B)
name = 'test_dataframe-9160c61b'

    def __init__(
        self,
        name: Optional[str] = None,
        *,
        software: str = None,
        n_workers: int = 4,
        worker_class: str = None,
        worker_options: dict = None,
        worker_vm_types: Optional[list] = None,
        worker_cpu: Union[int, List[int]] = None,
        worker_memory: Union[str, List[str]] = None,
        worker_disk_size: Optional[int] = None,
        worker_gpu: int = None,
        worker_gpu_type: Optional[str] = None,
        scheduler_class: str = None,
        scheduler_options: dict = None,
        scheduler_vm_types: Optional[list] = None,
        scheduler_cpu: Union[int, List[int]] = None,
        scheduler_memory: Union[str, List[str]] = None,
        asynchronous: bool = False,
        cloud: CloudBeta = None,
        account: str = None,
        shutdown_on_close=None,
        use_scheduler_public_ip: Optional[bool] = None,
        credentials: Optional[str] = "local",
        timeout: Optional[Union[int, float]] = None,
        environ: Optional[Dict[str, str]] = None,
        tags: Optional[Dict[str, str]] = None,
        send_dask_config: bool = True,
        backend_options: Optional[
            Union[AWSOptions, GCPOptions]
        ] = None,  # intentionally not in the docstring yet
        show_widget: bool = True,
        configure_logging: bool = False,
        wait_for_workers=None,
        package_sync: Union[bool, List[str]] = False,
        package_sync_strict: bool = False,
        package_sync_fail_on: Literal[
            "critical-only", "warning-or-higher", "any"
        ] = "critical-only",
        # easier network config
        allow_ingress_from: Optional[str] = None,
        allow_ssh: Optional[bool] = None,
    ):
        type(self)._instances.add(self)
        self.package_sync = bool(
            package_sync
        )  # TODO: this should be mutually exclusive with passing a software env
        if isinstance(package_sync, list):
            # ensure python is always included
            self.package_sync_only = set(package_sync)
            self.package_sync_only.add("python")
        else:
            self.package_sync_only = None
        self.package_sync_strict = package_sync_strict
        self.package_sync_fail_on = BEHAVIOR_TO_LEVEL[package_sync_fail_on]
        self.show_widget = show_widget
        self._cluster_status_logs = []
        if configure_logging:
            setup_logging()
    
        # Determine consistent sync/async
        if cloud and asynchronous is not None and cloud.asynchronous != asynchronous:
            warnings.warn(
                f"Requested a Cluster with asynchronous={asynchronous}, but "
                f"cloud.asynchronous={cloud.asynchronous}, so the cluster will be"
                f"{cloud.asynchronous}"
            )
    
            asynchronous = cloud.asynchronous
    
        self.scheduler_comm: Optional[dask.distributed.rpc] = None
    
        # It's annoying that the user must pass in `asynchronous=True` to get an async Cluster object
        # But I can't think of a good alternative right now.
        self.cloud: CloudBeta[IsAsynchronous] = cloud or CloudBeta.current(
            asynchronous=asynchronous
        )
        # if cloud:
        #     self.cleanup_cloud = False
        #     self.cloud: CloudBeta[IsAsynchronous] = cloud
        # else:
        #     self.cleanup_cloud = True
        #     self.cloud: CloudBeta[IsAsynchronous] = CloudBeta(asynchronous=asynchronous)
    
        # As of distributed 2021.12.0, deploy.Cluster has a ``loop`` attribute on the
        # base class. We add the attribute manually here for backwards compatibility.
        # TODO: if/when we set the minimum distributed version to be >= 2021.12.0,
        # remove this check.
        if DISTRIBUTED_VERSION >= "2021.12.0":
            kwargs = {"loop": self.cloud.loop}
        else:
            kwargs = {}
            self.loop = self.cloud.loop
    
        # we really need to call this first before any of the below code errors
        # out; otherwise because of the fact that this object inherits from
        # deploy.Cloud __del__ (and perhaps __repr__) will have AttributeErrors
        # because the gc will run and attributes like `.status` and
        # `.scheduler_comm` will not have been assigned to the object's instance
        # yet
        super(Cluster, self).__init__(asynchronous, **kwargs)
    
        self.timeout = (
            timeout if timeout is not None else self.cloud.default_cluster_timeout
        )
        self.software_environment = software or dask.config.get("coiled.software")
        self.worker_class = worker_class or dask.config.get("coiled.worker.class")
        self.worker_cpu = worker_cpu or dask.config.get("coiled.worker.cpu")
    
        if isinstance(worker_cpu, int) and worker_cpu <= 1:
            raise ValueError("`worker_cpu` should be at least 2.")
    
        self.worker_memory = worker_memory or dask.config.get("coiled.worker.memory")
        self.worker_vm_types = worker_vm_types
        self.worker_disk_size = worker_disk_size
        self.worker_gpu_count = worker_gpu
        self.worker_gpu_type = worker_gpu_type
        self.worker_options = {
            **(dask.config.get("coiled.worker-options", {})),
            **(worker_options or {}),
        }
    
        self.scheduler_class = scheduler_class or dask.config.get(
            "coiled.scheduler.class"
        )
    
        self.scheduler_class = scheduler_class or dask.config.get(
            "coiled.scheduler.class"
        )
    
        self.scheduler_class = scheduler_class or dask.config.get(
            "coiled.scheduler.class"
        )
        self.scheduler_cpu = scheduler_cpu or dask.config.get("coiled.scheduler.cpu")
        self.scheduler_memory = scheduler_memory or dask.config.get(
            "coiled.scheduler.memory"
        )
        self.scheduler_vm_types = scheduler_vm_types
        self.scheduler_options = {
            **(dask.config.get("coiled.scheduler-options", {})),
            **(scheduler_options or {}),
        }
    
        self.name = name or dask.config.get("coiled.name")
        self.account = account
        self._start_n_workers = n_workers
        self._lock = None
        self._asynchronous = asynchronous
        if shutdown_on_close is None:
            shutdown_on_close = dask.config.get("coiled.shutdown-on-close")
        self.shutdown_on_close = shutdown_on_close
        self.environ = {k: str(v) for (k, v) in (environ or {}).items() if v}
        self.tags = {k: str(v) for (k, v) in (tags or {}).items() if v}
        self.frozen_dask_config = (
            deepcopy(dask.config.config) if send_dask_config else {}
        )
        self.credentials = CredentialsPreferred(credentials)
        self._default_protocol = dask.config.get("coiled.protocol", "tls")
        self._wait_for_workers_arg = wait_for_workers
        self._last_logged_state_summary = None
    
        self.worker_vm_types = worker_vm_types
        self.scheduler_vm_types = scheduler_vm_types
    
        # these are sets of names of workers, only including workers in states that might eventually reach
        # a "started" state
        # they're used in our implementation of scale up/down (mostly inherited from coiled.Cluster)
        # and their corresponding properties are used in adaptive scaling (at least once we
        # make adaptive work with ClusterBeta).
        #
        # (Adaptive expects attributes `requested` and `plan`, which we implement as properties below.)
        #
        # Some good places to learn about adaptive:
        # https://github.com/dask/distributed/blob/39024291e429d983d7b73064c209701b68f41f71/distributed/deploy/adaptive_core.py#L31-L43
        # https://github.com/dask/distributed/issues/5080
        self._requested: Set[str] = set()
        self._plan: Set[str] = set()
    
        self._adaptive_options: Dict[str, Union[str, int]] = {}
        self.cluster_id: Optional[int] = None
        self.use_scheduler_public_ip: bool = (
            dask.config.get("coiled.use_scheduler_public_ip", True)
            if use_scheduler_public_ip is None
            else use_scheduler_public_ip
        )
    
        self.backend_options = backend_options
    
        if allow_ingress_from is not None or allow_ssh is not None:
            if backend_options is not None and "ingress" in backend_options:
                raise ArgumentCombinationError(
                    "You cannot use `allow_ingress_from` or `allow_ssh` when "
                    "`ingress` is also specified in `backend_options`."
                )
            firewall_kwargs = {
                "target": allow_ingress_from or "everyone",
                "ssh": False if allow_ssh is None else allow_ssh,
            }
    
            self.backend_options = self.backend_options or {}
            self.backend_options["ingress"] = cluster_firewall(**firewall_kwargs)[  # type: ignore
                "ingress"
            ]  # type: ignore
    
        if not self.asynchronous:
            # If we don't close the cluster, the user's ipython session gets spammed with
            # messages from distributed.
            #
            # Note that this doesn't solve all such spammy dead clusters (which is probably still
            # a problem), just spam created by clusters who failed initial creation.
            error = None
            try:
                self.sync(self._start)
            except (ClusterCreationError, InstanceTypeError) as e:
                error = e
                self.close()
                if self.cluster_id:
                    log_cluster_debug_info(self.cluster_id, self.account)
>               raise e.with_traceback(None)
E               coiled.errors.ClusterCreationError: Cluster status is error (reason: Scheduler Stopped -> Instance failed: Instance became unhealthy: State - shutting-down Reason - Server.InternalError: Internal error on launch) (cluster_id: 97224)

/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/coiled/_beta/cluster.py:413: ClusterCreationError
---------------------------- Captured stdout setup -----------------------------
Collecting git+https://github.com/dask/dask@43fede5413e9fad1499880a09e3d823a8b386a15
  Cloning https://github.com/dask/dask (to revision 43fede5413e9fad1499880a09e3d823a8b386a15) to /tmp/pip-req-build-rovwrlu5
Collecting git+https://github.com/dask/distributed@ec00cd590fbead781fa6f3960fd6f90c6155166e
  Cloning https://github.com/dask/distributed (to revision ec00cd590fbead781fa6f3960fd6f90c6155166e) to /tmp/pip-req-build-r1tafnxj
  Resolved https://github.com/dask/distributed to commit ec00cd590fbead781fa6f3960fd6f90c6155166e
  Preparing metadata (setup.py): started
  Resolved https://github.com/dask/dask to commit 43fede5413e9fad1499880a09e3d823a8b386a15
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Building wheels for collected packages: distributed
  Building wheel for distributed (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Building wheels for collected packages: dask
  Building wheel for dask (setup.py): started
  Building wheel for distributed (setup.py): finished with status 'done'
  Created wheel for distributed: filename=distributed-2022.10.0+9.gec00cd59-py3-none-any.whl size=911516 sha256=8751820d06c2e3f86e16ff889de8871d3e1224de0c8023adb10139224e9e8d26
  Stored in directory: /home/runner/.cache/pip/wheels/8f/16/34/a9b0dc103cd38d53d6805c32390703dedb64e281c83aa544d5
Successfully built distributed
  Building wheel for dask (setup.py): finished with status 'done'
  Created wheel for dask: filename=dask-2022.10.0+5.g43fede54-py3-none-any.whl size=1133847 sha256=66fac2565f2ef8574855109f44157a3760f4383f903470b2dc70a9df91c07616
  Stored in directory: /home/runner/.cache/pip/wheels/e4/2a/4b/d0f620af4f9c4eeb2b8a6f73024a3f3a1dfd74daa219b05778
Successfully built dask
╭───────────────────────────────────────── Package Issues ─────────────────────────────────────────╮
│               ╷                                                                   ╷              │
│   Package     │ Issue                                                             │ Risk Level   │
│ ╶─────────────┼───────────────────────────────────────────────────────────────────┼────────────╴ │
│   distributed │ Wheel built from                                                  │ Critical     │
│               │ git+[https://github.com/dask/distributed@ec00cd590fbead781fa6f396…](https://github.com/dask/distributed@ec00cd590fbead781fa6f396%E2%80%A6) │              │
│   dask        │ Wheel built from                                                  │ Critical     │
│               │ git+[https://github.com/dask/dask@43fede5413e9fad1499880a09e3d823…](https://github.com/dask/dask@43fede5413e9fad1499880a09e3d823%E2%80%A6) │              │
│   openssl     │ Package ignored                                                   │              │
│   abseil-cpp  │ Package ignored                                                   │              │
│               ╵                                                                   ╵              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
---------------------------- Captured stderr setup -----------------------------
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
DEPRECATION: --no-binary currently disables reading from the cache of locally built wheels. In the future --no-binary will not influence the wheel cache. pip 23.1 will enforce this behaviour change. A possible replacement is to use the --no-cache-dir option. You can use the flag --use-feature=no-binary-enable-wheel-cache to test the upcoming behaviour. Discussion can be found at https://github.com/pypa/pip/issues/11453
  Running command git clone --filter=blob:none --quiet https://github.com/dask/dask /tmp/pip-req-build-rovwrlu5
DEPRECATION: --no-binary currently disables reading from the cache of locally built wheels. In the future --no-binary will not influence the wheel cache. pip 23.1 will enforce this behaviour change. A possible replacement is to use the --no-cache-dir option. You can use the flag --use-feature=no-binary-enable-wheel-cache to test the upcoming behaviour. Discussion can be found at https://github.com/pypa/pip/issues/11453
  Running command git clone --filter=blob:none --quiet https://github.com/dask/distributed /tmp/pip-req-build-r1tafnxj
  Running command git rev-parse -q --verify 'sha^ec00cd590fbead781fa6f3960fd6f90c6155166e'
  Running command git fetch -q https://github.com/dask/distributed ec00cd590fbead781fa6f3960fd6f90c6155166e
  Running command git rev-parse -q --verify 'sha^43fede5413e9fad1499880a09e3d823a8b386a15'
  Running command git fetch -q https://github.com/dask/dask 43fede5413e9fad1499880a09e3d823a8b386a15
------------------------------ Captured log setup ------------------------------
INFO     coiled:cluster.py:623 Resolving your local python environment...
INFO     coiled.package_sync:magic.py:233 Conda environment detected: test
INFO     coiled.package_sync:magic.py:379 Attempting to create a wheel for dask in directory git+https://github.com/dask/dask@43fede5413e9fad1499880a09e3d823a8b386a15
INFO     coiled.package_sync:magic.py:379 Attempting to create a wheel for distributed in directory git+https://github.com/dask/distributed@ec00cd590fbead781fa6f3960fd6f90c6155166e
INFO     coiled.package_sync:magic.py:134 Loading conda metadata.json for https://conda.anaconda.org/conda-forge/linux-64
INFO     coiled.package_sync:magic.py:154 Downloading fresh conda repodata for https://conda.anaconda.org/conda-forge/linux-64
CRITICAL coiled:cluster.py:678 Dropped Package - dask, Wheel built from git+https://github.com/dask/dask@43fede5413e9fad1499880a09e3d823a8b386a15
CRITICAL coiled:cluster.py:678 Dropped Package - distributed, Wheel built from git+https://github.com/dask/distributed@ec00cd590fbead781fa6f3960fd6f90c6155166e
INFO     coiled:cluster.py:678 Dropped Package - openssl, Package ignored
INFO     coiled:cluster.py:678 Dropped Package - abseil-cpp, Package ignored
INFO     coiled:cluster.py:695 Environment magic complete
INFO     coiled:cluster.py:760 Creating Cluster (name: test_dataframe-9160c61b, https://cloud.coiled.io/dask-engineering/clusters/97224/details ). This might take a few minutes...
INFO     coiled:cluster.py:875 Ensuring network infrastructure is ready...
INFO     coiled:cluster.py:875 Scheduler: queued for launch                   Workers: 10 queued for launch (of 10)
ERROR    coiled:states.py:134    | Scheduler Instance     | test_dataframe-9160c61b-scheduler-01df006059-instance | error      at 17:25:21 (UTC) | Instance became unhealthy: State - shutting-down Reason - Server.InternalError: Internal error on launch
INFO     coiled:cluster.py:875 Scheduler: starting                            Workers: 10 queued for launch (of 10)
ERROR    coiled:states.py:134    | Scheduler Process      | test_dataframe-9160c61b-scheduler-01df006059   | error      at 17:25:25 (UTC) | Instance failed: Instance became unhealthy: State - shutting-down Reason - Server.InternalError: Internal error on launch
ERROR    coiled:states.py:134    | Cluster                | test_dataframe-9160c61b                        | error      at 17:25:26 (UTC) | Scheduler Stopped -> Instance failed: Instance became unhealthy: State - shutting-down Reason - Server.InternalError: Internal error on launch
INFO     coiled:cluster.py:875 Scheduler: error                               Workers: 10 queued for launch (of 10)
INFO     coiled:cluster.py:1030 Scheduler: error                               Workers: 10 queued for launch (of 10)
INFO     coiled:core.py:813 Cluster 97224 deleted successfully.
```

cc @shughes-uk @ntabris as it looks to be an internal `coiled`-related issue. I'll restart CI to see if it was just a transient issue

**Comment by ntabris** [comment]
deploy broke package sync, sorry, fix to be deployed later this afternoon

**Review by jrbourbeau — APPROVED** [review]

**Pull request merged**